### PR TITLE
Allow adding horse armor without enum magic

### PR DESCRIPTION
**Pull request by Victorious3**
I haven't tested this with the test mod yet (since I have no idea how to do that) but at least it works with vanilla. `EnumUtils` uses the sun package, which is a no go in any regards, and as you can see here, patching it to work without doesn't require much changes.

I have no idea how this registry system works, I'm going to change it for lazy loading, so just tell me if I did anything wrong.

### CODE REVIEW

**Comment by darkhax** [comment]
I am concerned with how stable this will be in the future, but if it works, it works.

**Comment by Victorious3** [comment]
@darkhax If they decide to change it, there's still the option to revert it back to an enum add a `CUSTOM` type + patch the methods (like you are already doing). They don't care about efficiency so there are always going to be 10 layers of indirection to exploit...

**Comment by shadowfacts** [comment]
I'm all for not using weird enum hacks, but I'm concerned about the maintainability of this. Vanilla uses enums quite frequently, often for things modders need access to, such as tool materials, armor materials, minecart types, etc.

**Comment by darkhax** [comment]
As @coolsquid has shown, patching every enum modders want access to is not a viable solution. We may need to undo this PR, depending on how further discussion goes.

**Comment by modmuss50** [comment]
The enum hacks thing can break, if mojang changes it

File: patches/net/minecraft/entity/passive/HorseArmorType.java.patch [hunk 1]
```diff
@@ -1,20 +1,61 @@
...
+-    IRON(5, "iron", "meo"),
+-    GOLD(7, "gold", "goo"),
+-    DIAMOND(11, "diamond", "dio");
+-
++public class HorseArmorType {
++    public static final HorseArmorType NONE = new HorseArmorType(0),
++    IRON = new HorseArmorType(5, "iron", "meo"),
++    GOLD = new HorseArmorType(7, "gold", "goo"),
++    DIAMOND = new HorseArmorType(11, "diamond", "dio");
++
```
Comment (phase) [inline patches/net/minecraft/entity/passive/HorseArmorType.java.patch]: unneeded whitespace here

**Comment by coolsquid** [comment]
> The enum hacks thing can break, if mojang changes it

Mojang can't do anything to break the enum hack, short of messing around with sun.reflect.Reflection (they won't). Forge has been using a similar thing for 4.5 years without issues.

Patching every enum class that modders might want to add to isn't feasible. EnumDyeColor alone would require us to patch nearly 30 classes. EnumUtils is way more stable, and far easier to maintain than hundreds of patches.

File: patches/net/minecraft/entity/passive/HorseArmorType.java.patch (hunk 1 above)
Comment (Victorious3) [inline patches/net/minecraft/entity/passive/HorseArmorType.java.patch]: +- 0, its already been one line each before

**Comment by Victorious3** [comment]
@coolsquid Most of these emus are downright useless, why do you have to patch dye colors? There's already a way to render things in any color you can possibly imagine. I don't think there are that many that need to be patched.

**Comment by Victorious3** [comment]
To clarify, feel free to do this search: https://github.com/MinecraftForge/MinecraftForge/search?utf8=%E2%9C%93&q=EnumHelper

See how often its actually used in practice. Of course this does not include the times where its beyond API usage so you might want to add the class back as a way to bypass, but I wouldn't use it for anything on our side.

File: patches/net/minecraft/entity/passive/HorseArmorType.java.patch (hunk 1 above)
Comment (phase) [inline patches/net/minecraft/entity/passive/HorseArmorType.java.patch]: Oh no, I thought I saw some spaces on that line.

**Comment by darkhax** [comment]
There are a lot of enums being used, you need to use it if you want to do any of the following
- Create a new tool tier (sword, pick, axe and so on)
- Create a new armor set (that doesn't have the same stats as vanilla)
- Create a new enchantment type (used by mods that want to sort their enchantments)
- Create a new banner pattern (Additional banners is the only one I know that does this)
- Add a new creature type (more important for mob mods)

Taking this small example of valid use cases, we would need at least 5 new patches.

**Comment by Victorious3** [comment]
@darkhax I'd call that a fair amount, and patching them to make them actually usable (new creature type? Not really, there's at least one switch/case that stops you from using it properly) is more beneficial to everyone. Having this be done through runtime magic only discourages people on working on a good API to use any of these.

**Pull request merged**